### PR TITLE
CDAP-4388 fix resource balancer to set completion

### PR DESCRIPTION
**Pull request by gsps1**
when finished is called on ResourceBalancerService#ResourceHandler, we stop the service (metrics.processor/log.saver) however we don't set completion on ResourceBalancer, causing it to continue running even after the service is stopped. By setting completion, this would stop resource balancer and other services in the metrics.processor/log.saver container when there is an exception or notifications are cancelled causing finished to be called.

JIRA : https://issues.cask.co/browse/CDAP-4388
Build : http://builds.cask.co/browse/CDAP-DUT3568-1

### CODE REVIEW

**Comment by anwar6953** [comment]
This doesn't resolve the root cause of https://issues.cask.co/browse/CDAP-4388, right?
What it does allow is for metrics processor to restart if the root cause is triggered, instead of just hanging in an idle state, right?

**Comment by gsps1** [comment]
yeah, also we will get the reason why the shutdown was triggered, as currently we don't log the cause. it will help to fix the issue if we see it in the future.

**Comment by anwar6953** [comment]
Ok, just keep in mind we shouldn't close the JIRA saying that this is the fix. Instead, this will allow insight in case it happens again.
Could possibly close the JIRA as cannot reproduce.

File: cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java [hunk 1]
```diff
@@ -195,8 +195,21 @@ public void onChange(Collection<PartitionReplica> partitionReplicas) {
...
+              if (failureCause != null) {
+                LOG.error("No more changes will be notified to the service: {}.", serviceName, failureCause);
+                completion.setException(failureCause);
+              } else {
+                LOG.info("No more changes will be notified to the service: {}.", serviceName);
+                completion.set(null);
+              }
+            }
+          } else {
+            completion.set(null);
```
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java]: I don't see how the `completion` is being used in the code. Do you know how it is being used? All I see is setting the state of the `completion` but no one is trying to get it.

Comment (gsps1) [inline cdap-common/src/main/java/co/cask/cdap/common/resource/ResourceBalancerService.java]: yeah, you are right, it seems like this completion future variable is not actually needed, based on the logic in ResourceHandler, we would want to restart metrics/log.saver services on partition change's and we would not want to stop the resource balancer service on those scenarios as well as  exception during starting/stopping metrics/log.saver service. Its not clear to me when the ``ResourceBalancer#shutdown`` would be called. 
But i looked at the logs again for the issue, and it's clear the ``onChange`` was called twice(both happened on the same timestamp), first time with non-empty partitions list and the second time with an empty list, the first call started the metrics.processor service, while the second call stopped the service, which is expected as we stop the service during ``onChange`` and don't start it back if the list is empty.  However its not clear why an empty partition's list  was sent..

**Comment by gsps1** [comment]
will close this PR, this doesn't fix the issue in JIRA, we can remove usage of completion as its not useful, but that will just be cleanup. next step : investigate why onChange was called with empty partitions and try to reproduce that.

**Comment by anwar6953** [comment]
We should at least include the logging that was happening in this PR,
right? That will help debug the issue if it happens again.

**Comment by gsps1** [comment]
the partitions became empty and that led to stopping the service, logging wouldn't help in this case.

**Comment by anwar6953** [comment]
I see. Basically, what you're saying is that we know the reason (failureCause)?

**Pull request closed without merging**